django/db/models/fields/related.py

```diff
@@ -731,26 +732,15 @@ def get_reverse_path_info(self):
         pathinfos = [PathInfo(from_opts, opts, (opts.pk,), self.remote_field, not self.unique, False)]
         return pathinfos
 
-    def get_lookup(self, lookup_name):
-        if lookup_name == 'in':
-            return RelatedIn
-        elif lookup_name == 'exact':
-            return RelatedExact
-        elif lookup_name == 'gt':
-            return RelatedGreaterThan
-        elif lookup_name == 'gte':
-            return RelatedGreaterThanOrEqual
-        elif lookup_name == 'lt':
-            return RelatedLessThan
-        elif lookup_name == 'lte':
-            return RelatedLessThanOrEqual
-        elif lookup_name == 'isnull':
-            return RelatedIsNull
-        else:
-            raise TypeError('Related Field got invalid lookup: %s' % lookup_name)
+    @classmethod
+    def get_lookups(cls):
+        if 'cached_lookups' not in cls.__dict__:
```

Comment:
The caching here is based on the same idea as `class_lookups`, just a bit more complicated because of the cache busting. I didn't try `@cached_property` (no way to bust the cached value, and it doesn't work w/ classmethods, right?), but `lru_cache` worked just fine. That said, it is about 50% slower.
  - current implementation at 100k runs takes ~.28s on average.
  - `lru_cache` at 100k runs takes ~.42s on average.

50% is a significant percentage, but we're talking about lookup resolution taking .0000042 seconds vs .0000028 seconds. `¯\_(ツ)_/¯`

Either way, I've squashed the existing commits into one, and added the lru changes separately. Thoughts on the performance vs code clarity?